### PR TITLE
feat(Profile): New Social Links workflow

### LINKED ISSUE #9777: New Social Links workflow: all links visible, order of items saved, drag'n'drop, editing/deleting in a popup, several links of the same type

### PR DESCRIPTION
**Pull request by caybro**
TLDR;
- all links visible in settings/popup
- persistent order of items
- drag'n'drop to reorder
- editing/deleting in a new popup
- several links of the same type

Added a generic StatusDraggableItem that will be reused later for other parts of the Profile Showcase (assets/collectibles/etc)

Removed now unused components

Needs changes in nimqml (to expose QAIM::begin/endMoveRows), in DOtherSide (to expose those to NIM), in status-go (to preserve the links order and fully save them to DB)

Needs: https://github.com/status-im/nimqml/pull/49
Needs: https://github.com/status-im/status-go/pull/3276

Fixes #9777

### What does the PR do

Implements a new, full end-to-end workflow to Social Links (part of the Profile Showcase feature)

### Affected areas

Settings/Profile

### StatusQ checklist

- [x] add documentation if necessary (new component, new feature)
- [x] update sandbox app
  - in case of new component, add new component page
  - in case of new features, add variation to existing component page
  - nice to have: add it to the demo application as well
- [x] test changes in both light and dark theme?

### Screenshot of functionality (including design for comparison)

- [x] I've checked the design and this PR matches it

Empty links:
image

Adding a link:
image
image
image
image

Editing a link:
image

Changed order (Telegram first):
image
image

Profile popup shows the same links (with correct order):
image

### CODE REVIEW

**Review by noeliaSD — COMMENTED** [review]
Very good job!! It looks nice! I've just played with that, no code review at that time (will do it tomorrow)

1. Here some questions related to diferences between the app and the design I can see:
- Should "Your links will be here" box from design be displayed if the user has no social links?
- Should "Communities here will show on your profile" box from design be displayed if the user has no communities?
- Should the tabs `Communities / Accounts / Collectibles / Assets` be shown always although they are empty?
Screenshot 2023-03-14 at 15 47 34

2. Maybe it's not related to this pr but I'm really having trouble scrolling in this view... If I have my mouse on the links area... the scrolling doesn't work and I need to go to a small area to move up and down. See video for a better understanding:

https://user-images.githubusercontent.com/97019400/225064147-077d9962-fd5f-4b6e-a272-659e1fbe4315.mov

**Comment by caybro** [comment]
> Very good job!! It looks nice! I've just played with that, no code review at that time (will do it tomorrow)
> 
> 1. Here some questions related to diferences between the app and the design I can see:
> 
> * Should "Your links will be here" box from design be displayed if the user has no social links?

Yes, is it not the case for you? (here it works fine)

> * Should "Communities here will show on your profile" box from design be displayed if the user has no communities?
> * Should the tabs `Communities / Accounts / Collectibles / Assets` be shown always although they are empty?
>   Screenshot 2023-03-14 at 15 47 34

Those bottom tabs are not handled in this PR; just the Social Links section

> 
> 2. Maybe it's not related to this pr but I'm really having trouble scrolling in this view... If I have my mouse on the links area... the scrolling doesn't work and I need to go to a small area to move up and down. See video for a better understanding:

Yeah, that's unfortunately a clash between the Flickable that handles the scroll events, and the list items that need the mouse grab in order to perform the drag'n'drop; I fear there's nothing that could fix this

**Comment by noeliaSD** [comment]
> > 
> > * Should "Your links will be here" box from design be displayed if the user has no social links?
> 
> Yes, is it not the case for you? (here it works fine, see the first screenshot in this PR)
> 

No, I can't see the box when there are no social links. See the pic:

Screenshot 2023-03-15 at 08 47 25

**Comment by noeliaSD** [comment]
> > 2. Maybe it's not related to this pr but I'm really having trouble scrolling in this view... If I have my mouse on the links area... the scrolling doesn't work and I need to go to a small area to move up and down. See video for a better understanding:
> 
> Yeah, that's unfortunately a clash between the Flickable that handles the scroll events, and the list items that need the mouse grab in order to perform the drag'n'drop; I fear there's nothing that could fix this

I'm wondering if it would make sense to restrict the drag'n drop just to the area marked below so, the complete item in not clickable and maybe we have more are to do the scrolling?!

Screenshot 2023-03-15 at 08 51 56

File: ui/StatusQ/src/StatusQ/Components/StatusDraggableListItem.qml [hunk 1]
```diff
@@ -0,0 +1,116 @@
+import QtQuick 2.15
+import QtQuick.Controls 2.15
+import QtQuick.Layouts 1.15
+
+import StatusQ.Core 0.1
+import StatusQ.Core.Theme 0.1
+
+ItemDelegate {
```
Comment (noeliaSD) [inline ui/StatusQ/src/StatusQ/Components/StatusDraggableListItem.qml:8]: It would be nice if you write documentation!!

File: ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+                height: draggableDelegate.height
+
+                keys: ["x-status-draggable-list-item-internal"]
+
+                onEntered: function(drag) {
+                    const from = drag.source.visualIndex
+                    const to = draggableDelegate.visualIndex
+                    if (to === from)
+                        return
+                    console.debug("!!! MOVE FROM", from, "TO", to)
```
Comment (noeliaSD) [inline ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml:163]: leftover?

File: ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+                    root.profileStore.moveLink(from, to, 1)
+                    drag.accept()
+                }
+
+                onDropped: (drop) => console.debug("!!! DROPPED AT", drop.source.visualIndex)
```
Comment (noeliaSD) [inline ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml:168]: leftover?

File: ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+                text: qsTr("＋ Add more links")
+                color: Theme.palette.primaryColor1
+                font.pixelSize: Theme.tertiaryTextFontSize
+                font.weight: Font.Normal
+                onClicked: Global.openPopup(addSocialLinkModalComponent)
+            }
+        }
+
+        // empty placeholder when no links; dashed rounded rectangle
+        Shape {
```
Comment (noeliaSD) [inline ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml:62]: Maybe you can make it a component in a different file. You will need it, at least, in empty communities, assets and so on tabs.. and probably it will be extended also to other views!

File: ui/app/AppLayouts/Profile/popups/AddSocialLinkModal.qml [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+import utils 1.0
+
+import StatusQ.Core 0.1
+import StatusQ.Core.Theme 0.1
+import StatusQ.Controls 0.1
+import StatusQ.Components 0.1
+import StatusQ.Popups 0.1
+
+import "../controls"
```
Comment (noeliaSD) [inline ui/app/AppLayouts/Profile/popups/AddSocialLinkModal.qml:13]:
```suggestion
import AppLayouts.Profile.controls 1.0
```

File: ui/app/AppLayouts/Profile/popups/AddSocialLinkModal.qml [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    padding: currentIndex === 0 ? 0 : Style.current.padding
+
+    header.title: qsTr("Add more links")
+    rightButtons: [finishButton]
+    finishButton: StatusButton {
+        text: qsTr("Add")
+        enabled: !!linkTarget.text
+        onClicked: {
+            root.addLinkRequested(d.selectedLinkTypeText || customTitle.text, // text for custom link, otherwise the link typeId
+                                  Utils.addSocialLinkPrefix(linkTarget.text, d.selectedLinkType),
```
Comment (noeliaSD) [inline ui/app/AppLayouts/Profile/popups/AddSocialLinkModal.qml:32]: Does it make sense to have it in `ProfileUtils` instead?

File: ui/app/AppLayouts/Profile/popups/ModifySocialLinkModal.qml [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+import utils 1.0
+
+import StatusQ.Core 0.1
+import StatusQ.Core.Theme 0.1
+import StatusQ.Controls 0.1
+import StatusQ.Components 0.1
+import StatusQ.Popups.Dialog 0.1
+
+import "../controls"
```
Comment (noeliaSD) [inline ui/app/AppLayouts/Profile/popups/ModifySocialLinkModal.qml:14]:
```suggestion
import AppLayouts.Profile.controls 1.0
```

**Review by noeliaSD — COMMENTED** [review]
Added some minor comments / suggestions

File: ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml (hunk 2 above)
Comment (caybro) [reply ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml:163]: Well not really... it's a debug, so it won't appear in production builds; I can remove if you insist :)

File: ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml (hunk 3 above)
Comment (caybro) [reply ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml:168]: Same as above

File: ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml (hunk 4 above)
Comment (caybro) [reply ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml:62]: Yup, good idea!

File: ui/app/AppLayouts/Profile/popups/AddSocialLinkModal.qml (hunk 6 above)
Comment (caybro) [reply ui/app/AppLayouts/Profile/popups/AddSocialLinkModal.qml:32]: Yup, I meant to do it at one point but forgot, makes sense :)

**Comment by caybro** [comment]
> > > * Should "Your links will be here" box from design be displayed if the user has no social links?
> > 
> > 
> > Yes, is it not the case for you? (here it works fine, see the first screenshot in this PR)
> 
> No, I can't see the box when there are no social links. See the pic:
> 
> Screenshot 2023-03-15 at 08 47 25

Right, I guess I'll have to readd the proxy model to count the non-empty items

**Comment by caybro** [comment]
> > > 2. Maybe it's not related to this pr but I'm really having trouble scrolling in this view... If I have my mouse on the links area... the scrolling doesn't work and I need to go to a small area to move up and down. See video for a better understanding:
> > 
> > 
> > Yeah, that's unfortunately a clash between the Flickable that handles the scroll events, and the list items that need the mouse grab in order to perform the drag'n'drop; I fear there's nothing that could fix this
> 
> I'm wondering if it would make sense to restrict the drag'n drop just to the area marked below so, the complete item in not clickable and maybe we have more are to do the scrolling?!
> 
> Screenshot 2023-03-15 at 08 51 56

Hmmm... tough one; any idea @benjthayer @John-44 ?

File: ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml (hunk 2 above)
Comment (noeliaSD) [reply ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml:163]: Just a question! Up to you! ;)

Comment (caybro) [reply ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml]: Removed :)

File: ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml (hunk 4 above)
Comment (caybro) [reply ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml]: Done

File: ui/app/AppLayouts/Profile/popups/AddSocialLinkModal.qml (hunk 6 above)
Comment (caybro) [reply ui/app/AppLayouts/Profile/popups/AddSocialLinkModal.qml]: Added

**Comment by caybro** [comment]
@noeliaSD everything should be fixed/addressed (except the question whether the whole item should be draggable, or just the handle). Please give it another look

File: test/ui-test/testSuites/suite_settings/tst_userIdentity/test.feature [hunk 8]
```diff
@@ -9,6 +9,7 @@ Feature: User Identity
         And the user signs up with username "tester123" and password "TesTEr16843/!@00"
         And the user lands on the signed in app
 
+    @mayfail
```
Comment (noeliaSD) [inline test/ui-test/testSuites/suite_settings/tst_userIdentity/test.feature:12]: Please, remember to add a task for refactoring this test!!!

File: ui/StatusQ/src/StatusQ/Components/StatusDraggableListItem.qml [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+
+import StatusQ.Core 0.1
+import StatusQ.Core.Theme 0.1
+
+/*!
+   \qmltype StatusDraggableListItem
+   \inherits QtQuickControls::ItemDelegate
+   \inqmlmodule StatusQ.Components
+   \since StatusQ.Components 0.1
+   \brief It is a list item with ther ability to be dragged and dropped to reorder within a list view. Inherits from \c QtQuickControls::ItemDelegate.
```
Comment (noeliaSD) [inline ui/StatusQ/src/StatusQ/Components/StatusDraggableListItem.qml:13]:
```suggestion
   \brief It is a list item with the ability to be dragged and dropped to reorder within a list view. Inherits from \c QtQuickControls::ItemDelegate.
```

File: ui/StatusQ/src/StatusQ/Components/StatusDraggableListItem.qml (hunk 1 above)
Comment (noeliaSD) [reply ui/StatusQ/src/StatusQ/Components/StatusDraggableListItem.qml:98]: You can add it also here: `statusqcomponents.qdoc`

**Review by noeliaSD — APPROVED** [review]
Tested again! I can see the shape rectangle now! :)

Just added 2 minor comments, otherwise LGTM!

File: ui/StatusQ/src/StatusQ/Components/StatusDraggableListItem.qml (hunk 9 above)
Comment (caybro) [reply ui/StatusQ/src/StatusQ/Components/StatusDraggableListItem.qml:13]: Ok, will do in another push/PR

File: test/ui-test/testSuites/suite_settings/tst_userIdentity/test.feature (hunk 8 above)
Comment (caybro) [reply test/ui-test/testSuites/suite_settings/tst_userIdentity/test.feature:12]: Created/added under https://github.com/status-im/status-desktop/issues/9909

Comment (noeliaSD) [reply test/ui-test/testSuites/suite_settings/tst_userIdentity/test.feature:12]: Thanks!!

**Comment by caybro** [comment]
> Hmmm... tough one; any idea @benjthayer @John-44 ?

> If both options are equal dev effort, then I think draging anywhere on the whole item would be best, but if this is a challange to implement we could also do 'user can only drag using the drag handle' which would be fine

All good then :)

**Review by alexjba — COMMENTED** [review]
Nice work! 
Minor comments. Mainly on large text handling in Text boxes.

Also, might be worth looking at this since you're working on the social links. The tooltip can go crazy with large text:

Screenshot 2023-03-15 at 19 49 43

File: src/app/modules/main/profile_section/profile/view.nim [hunk 10]
```diff
@@ -90,28 +90,35 @@ QtObject:
...
-  proc updateLink(self: View, uuid: string, text: string, url: string): bool {.slot.} =
+  proc updateLink(self: View, uuid: string, text: string, url: string) {.slot.} =
     if (self.temporarySocialLinksModel.updateItem(uuid, text, url)):
       self.temporarySocialLinksJsonChanged()
       self.socialLinksDirtyChanged()
 
-  proc resetSocialLinks(self: View): bool {.slot.} =
+  proc moveLink(self: View, fromRow: int, toRow: int) {.slot.} =
+    if (self.temporarySocialLinksModel.moveItem(fromRow, toRow)):
+      let res = self.delegate.saveSocialLinks() # save silently
```
[the remark this answers is not in the document]
Comment (alexjba) [reply src/app/modules/main/profile_section/profile/view.nim:110]: https://user-images.githubusercontent.com/47811206/225408251-7570548e-c77b-42b3-9957-ac4293b53e8b.mov

**Comment by imagiMary** [comment]
Hi @caybro Looks good! Just one thing, if the users handle is really long the edit button disappears under neither the handle and then it becomes impossible for the user to edit that item. 

image

File: ui/app/AppLayouts/Profile/views/profile/MyProfileSettingsView.qml [hunk 11]
```diff
@@ -144,17 +119,22 @@ ColumnLayout {
...
 
-        onSocialLinkChanged: profileStore.updateLink(uuid, text, url)
-        onAddSocialLinksClicked: socialLinksModal.open()
+    Separator {
+        Layout.fillWidth: true
     }
 
-    SocialLinksModal {
-        id: socialLinksModal
+    ProfileSocialLinksPanel {
```
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Profile/views/profile/MyProfileSettingsView.qml:128]: No binding loops on my side (Qt 5.15.10)

File: ui/app/AppLayouts/Profile/popups/AddSocialLinkModal.qml [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+            }
+        },
+        ColumnLayout {
+            width: root.availableWidth
+            spacing: Style.current.halfPadding
+
+            StaticSocialLinkInput {
+                id: customTitle
+                Layout.fillWidth: true
+                visible: d.selectedLinkType === Constants.socialLinkType.custom
```
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Profile/popups/AddSocialLinkModal.qml:106]: Yup, I'll add the character limit to QML, it's set to `24` in status-go. It's true that it might silently fail saving but the only validation status-go does is checking for non-empty URL (we don't allow that in QML) and the above length (will add)

File: src/app/modules/main/profile_section/profile/view.nim (hunk 10 above)
[the remark this answers is not in the document]
Comment (caybro) [reply src/app/modules/main/profile_section/profile/view.nim:110]: Yeah... I was thinking of splitting the move/save into 2 operations; it doesn't freeze here but I can see we are probably saving needlessly too often. I'll give it a try

**Comment by caybro** [comment]
> Hi @caybro Looks good! Just one thing, if the users handle is really long the edit button disappears under neither the handle and then it becomes impossible for the user to edit that item.
> 
> image

Yup, will fix

File: ui/app/AppLayouts/Profile/popups/AddSocialLinkModal.qml (hunk 12 above)
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Profile/popups/AddSocialLinkModal.qml:106]: Added `Constants.maxSocialLinkTextLength: 24` and using it for the social link text fields (that's what the status-go validation does):

```go
const (
    maxSocialLinkTextLength = 24
)

func ValidateSocialLinks(socialLinks *identity.SocialLinks) error {
    for _, link := range *socialLinks {
        if len(link.Text) > maxSocialLinkTextLength {
            return ErrInvalidSocialLinkTextLength
        }
    }
    return nil
}
```

File: ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+                    dragParent: linksView
+                    visualIndex: delegateRoot.visualIndex
+                    draggable: linksView.count > 1
+                    title: ProfileUtils.linkTypeToText(model.linkType) || model.text
+                    icon.name: model.icon
+                    icon.color: ProfileUtils.linkTypeColor(model.linkType)
+                    actions: [
+                        StatusLinkText {
+                            font.pixelSize: Theme.primaryTextFontSize
+                            font.weight: Font.Normal
```
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Profile/panels/ProfileSocialLinksPanel.qml:139]: Fixed:

image

File: src/app/modules/main/profile_section/profile/view.nim (hunk 10 above)
[the remark this answers is not in the document]
Comment (caybro) [reply src/app/modules/main/profile_section/profile/view.nim]: Fixed (DND split into 2 parts; `moveRows` on pos changed, and `saveLinks` on drop)

File: ui/app/AppLayouts/Profile/popups/AddSocialLinkModal.qml (hunk 12 above)
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Profile/popups/AddSocialLinkModal.qml:106]: image

**Comment by caybro** [comment]
@alexjba all issues addressed/fixed, pls have another look, thx :)

**Review by alexjba — APPROVED** [review]
Looks great! Nice work! 👍 

I guess this issue is out of scope, right? If so we need to add a task for it. Here is an example with a valid but long URL:
Screenshot 2023-03-16 at 10 36 14

**Comment by caybro** [comment]
> Looks great! Nice work! +1
> 
> I guess this issue is out of scope, right? If so we need to add a task for it. Here is an example with a valid but long URL: Screenshot 2023-03-16 at 10 36 14

Yes pls, create a task for StatusTooltip, I'll fix it in a different PR
